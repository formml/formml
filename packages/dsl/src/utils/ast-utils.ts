import { AstNode } from 'langium'

import { Literal, NullLiteral, isNullLiteral } from '../language/index.js'

type LiteralValue<T extends Literal> = T extends { value: unknown }
  ? T['value']
  : T extends NullLiteral
    ? null
    : never

export function resolveLiteralValue<T extends Literal>(
  literal: T,
): LiteralValue<T>
export function resolveLiteralValue(literal: Literal) {
  if (isNullLiteral(literal)) {
    return null
  }
  return literal.value
}

export function stringify(node: AstNode, space?: number | string): string {
  return JSON.stringify({ node }, null, space)
}
