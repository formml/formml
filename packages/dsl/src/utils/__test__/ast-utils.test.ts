import { resolveLiteralValue, stringify } from '../ast-utils.js'

describe('ast utils', () => {
  describe('resolveLiteralValue', () => {
    test('should return string value given string literal', () => {
      const literal = {
        $container: {} as never,
        $type: 'TextLiteral' as const,
        value: 'hello',
      }
      expect(resolveLiteralValue(literal)).toBe('hello')
    })

    test('should return number value given number literal', () => {
      const literal = {
        $container: {} as never,
        $type: 'NumLiteral' as const,
        value: 123,
      }
      expect(resolveLiteralValue(literal)).toBe(123)
    })

    test('should return boolean value given boolean literal', () => {
      const literal = {
        $container: {} as never,
        $type: 'BoolLiteral' as const,
        value: true,
      }
      expect(resolveLiteralValue(literal)).toBe(true)
    })

    test('should return null given null literal', () => {
      const literal = {
        $container: {} as never,
        $type: 'NullLiteral' as const,
      }
      expect(resolveLiteralValue(literal)).toBe(null)
    })
  })

  describe('stringify', () => {
    test('should return json string given a simple ast node', () => {
      const node = {
        $type: 'SimpleNode',
        propA: 'valueA',
        propB: 123,
        propC: true,
      }

      expect(stringify(node)).toMatchInlineSnapshot(
        `"{"node":{"$type":"SimpleNode","propA":"valueA","propB":123,"propC":true}}"`,
      )
    })

    test.each([2, '[tab]'])('should accept space parameter', (space) => {
      const node = {
        $type: 'SimpleNode',
        propA: 'valueA',
        propB: 123,
        propC: true,
      }

      expect(stringify(node, space)).toMatchSnapshot()
    })
  })
})
