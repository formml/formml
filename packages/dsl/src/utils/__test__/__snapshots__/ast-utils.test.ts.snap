// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ast utils > stringify > should accept space parameter 1`] = `
"{
  "node": {
    "$type": "SimpleNode",
    "propA": "valueA",
    "propB": 123,
    "propC": true
  }
}"
`;

exports[`ast utils > stringify > should accept space parameter 2`] = `
"{
[tab]"node": {
[tab][tab]"$type": "SimpleNode",
[tab][tab]"propA": "valueA",
[tab][tab]"propB": 123,
[tab][tab]"propC": true
[tab]}
}"
`;
